FIAT/pointwise_dual.py

```diff
@@ -0,0 +1,52 @@
+# Copyright (C) 2020 Robert C. Kirby (Baylor University)
+#
+# This file is part of FIAT (https://www.fenicsproject.org)
+#
+# SPDX-License-Identifier:    LGPL-3.0-or-later
+#
+import numpy as np
+from FIAT.functional import Functional
+from FIAT.dual_set import DualSet
+
+
+def compute_pointwise_dual(el, pts):
+    """Constructs a dual basis to the basis for el as a linear combination
+    of a set of pointwise evaluations.  This is useful when the
+    prescribed finite element isn't Ciarlet (e.g. the basis functions
+    are provided explicitly as formulae).  Alternately, the element's
+    given dual basis may involve differentiation, making run-time
+    interpolation difficult in FIAT clients.  The pointwise dual,
+    consisting only of pointwise evaluations, will effectively replace
+    these derivatives with (automatically determined) finite
+    differences.  This is exact on the polynomial space, but is an
+    approximation if applied to functions outside the space.
+
+    :param el: a :class:`FiniteElement`.
+    :param pts: an iterable of points with the same length as el's
+                dimension.  These points must be unisolvent for the
+                polynomial space
+    :returns: a :class `DualSet`
+    """
+    nbf = el.space_dimension()
+
+    T = el.ref_el
+    z = tuple([0] * T.get_dimension())
+
+    nds = []
+
+    V = el.tabulate(0, pts)[z]
+
+    alphas = np.linalg.inv(V.reshape((nbf, -1)).T).reshape(V.shape)
+    for _, coeffs in enumerate(alphas):
+        pt_dict = {}
+        for k in range(coeffs.shape[-1]):
+            lst = []
+            for comp in np.ndindex(coeffs.shape[:-1]):
+                blah = tuple(list(comp) + [k])
+                if np.abs(coeffs[blah]) >= 1.e-12:
+                    lst.append((coeffs[blah], comp))
+            if lst != []:
+                pt_dict[pts[k]] = lst
```

Comment:
All of this back and forth indexing is rather opaque. Can you add a comment describing what's happening? I think I already lost track of what's going where.

Reply:
Thinking of comments to add.  It's much cleaner in the scalar case, but I hate to keep a special case around for pedagogical reasons.  

We already have a bunch of opaque voodoo in FIAT, surely a little more won't hurt? ;)

Reply:
OK, see if current version makes it less bad.  It's pretty tight code -- not going to be easy to read.

Reply:
OK, I think I understand. This can be written slightly more array-like:
```python
from collections import defaultdict

...

pts = numpy.asarray(pts) 
for coeffs in alphas: 
    pt_dict = defaultdict(list) 
    where = numpy.where(numpy.abs(coeffs) > 1e-12) 
    *comp, pt_index = where 
    for pt, coeff_comp in zip(pts[pt_index], zip(coeffs[where], zip(*comp))): 
        pt_dict[tuple(pt)].append(coeff_comp) 
    pt_dict = dict(pt_dict) 
```

I think. Perhaps that is not more readable.

Reply:
More readable but not quite correct (fails in the all-important scalar case since the `comp` things comes up empty.  Trying to find a workaround that doesn't involve too much special casing.

Reply:
Ok, I punted and have an `if` statement inside the `for` loop.  Take a look and see if you can find a better way.  The problem is that in scalar cases, `*comp` comes back as an empty tuple so when you zip it against something you get an empty iterator and we never enter the loop that sticks things into the `pt_dict`.